File: backend/app/schemas/mission.py
import time
from typing import Literal

from pydantic import BaseModel, Field


MissionStage = Literal["stage1", "stage2", "stage3"]
Stage1Target = Literal["Balistik Füze", "Helikopter", "F16", "Mini/Micro İHA"]
Stage1RangeScore = Literal[5, 10, 20]
Stage3TargetClass = Literal["f16", "helicopter", "ballistic_missile", "mini_micro_uav"]

STAGE1_REQUIRED_TARGETS: tuple[Stage1Target, ...] = ("Balistik Füze", "Helikopter", "F16", "Mini/Micro İHA")


class Stage1HitRequest(BaseModel):
    target: Stage1Target
    score_awarded: Stage1RangeScore


class Stage1WrongTargetRequest(BaseModel):
    target: Stage1Target


class Stage1ManualMotionRequest(BaseModel):
    speed_x: int = Field(ge=-1000, le=1000)
    speed_y: int = Field(ge=-1000, le=1000)
    duration_ms: int = Field(ge=50, le=1200)


class Stage1PlanRequest(BaseModel):
    order: list[Stage1Target] = Field(min_length=4, max_length=4)


class Stage1Event(BaseModel):
    kind: Literal["hit", "wrong_target"]
    target: Stage1Target
    score_awarded: int = 0
    penalty: int = 0
    elapsed_s: int
    timestamp: float = Field(default_factory=time.time)


class Stage2RoundCompleteRequest(BaseModel):
    confirmed_hits: int = Field(ge=0, le=3)


class Stage2RoundEvent(BaseModel):
    round_number: int = Field(ge=1, le=4)
    confirmed_hits: int = Field(ge=0, le=3)
    points: int
    zero_hit_streak: int = Field(ge=0)
    timestamp: float = Field(default_factory=time.time)


class Stage3RoundCompleteRequest(BaseModel):
    enemy_class: Stage3TargetClass
    enemy_hit: bool
    friend_hit: bool = False


class Stage3RoundEvent(BaseModel):
    round_number: int = Field(ge=1, le=8)
    enemy_class: Stage3TargetClass
    enemy_hit: bool
    friend_hit: bool
    points: int
    penalty: int
    miss_streak: int = Field(ge=0)
    timestamp: float = Field(default_factory=time.time)


class MissionState(BaseModel):
    active_stage: MissionStage = "stage1"
    elapsed_s: int = 0
    timer_running: bool = False
    stage1_hits: int = 0
    stage1_wrong_hits: int = 0
    stage1_order: list[str] = Field(default_factory=lambda: ["Balistik Füze", "Helikopter", "F16", "Mini/Micro İHA"])
    stage1_order_locked: bool = False
    stage1_completed_targets: list[Stage1Target] = Field(default_factory=list)
    stage1_raw_points: int = 0
    stage1_penalty_points: int = 0
    stage1_events: list[Stage1Event] = Field(default_factory=list)
    stage2_round: int = 1
    stage2_hits: int = 0
    stage2_completed_rounds: int = 0
    stage2_round_events: list[Stage2RoundEvent] = Field(default_factory=list)
    stage2_zero_hit_streak: int = 0
    stage2_failed: bool = False
    stage3_round: int = 1
    stage3_hits: int = 0
    stage3_friend_or_miss_penalties: int = 0
    stage3_completed_rounds: int = 0
    stage3_round_events: list[Stage3RoundEvent] = Field(default_factory=list)
    stage3_miss_streak: int = 0
    stage3_failed: bool = False
    updated_at: float = Field(default_factory=time.time)


class MissionUpdate(BaseModel):
    active_stage: MissionStage | None = None
    elapsed_s: int | None = Field(default=None, ge=0)
    timer_running: bool | None = None
    stage1_hits: int | None = Field(default=None, ge=0)
    stage1_wrong_hits: int | None = Field(default=None, ge=0)
    stage1_order: list[str] | None = None
    # Kept only to return STAGE2_EVENT_API_REQUIRED to legacy clients rather
    # than silently accepting an obsolete mutable score update.
    stage2_round: int | None = Field(default=None, ge=1, le=4)
    stage2_hits: int | None = Field(default=None, ge=0)
    stage3_round: int | None = Field(default=None, ge=1, le=8)
    stage3_hits: int | None = Field(default=None, ge=0)
    stage3_friend_or_miss_penalties: int | None = Field(default=None, ge=0)


class MissionScore(BaseModel):
    stage1_score: int
    stage2_score: int
    stage3_score: int
    active_score: int
    remaining_s: int
    total_estimated_score: int
    stage1_raw_points: int = 0
    stage1_penalty_points: int = 0
    stage1_bonus_points: int = 0
    stage1_next_target: Stage1Target | None = None
    stage1_plan_locked: bool = False
    stage2_round_scores: list[int] = Field(default_factory=list)
    stage2_zero_hit_streak: int = 0
    stage2_failed: bool = False
    stage2_passing_threshold_met: bool = False
    stage3_round_scores: list[int] = Field(default_factory=list)
    stage3_miss_streak: int = 0
    stage3_failed: bool = False
    stage3_award_threshold_met: bool = False


class MissionSnapshot(BaseModel):
    state: MissionState
    score: MissionScore
    no_physical_command_generated: bool = True

File: frontend/src/types/mission.ts
export type MissionStage = 'stage1' | 'stage2' | 'stage3'
export type Stage1Target = 'Balistik Füze' | 'Helikopter' | 'F16' | 'Mini/Micro İHA'
export type Stage1RangeScore = 5 | 10 | 20
export type Stage3TargetClass = 'f16' | 'helicopter' | 'ballistic_missile' | 'mini_micro_uav'

export interface Stage1Event {
  kind: 'hit' | 'wrong_target'
  target: Stage1Target
  score_awarded: number
  penalty: number
  elapsed_s: number
  timestamp: number
}

export interface Stage2RoundEvent {
  round_number: number
  confirmed_hits: number
  points: number
  zero_hit_streak: number
  timestamp: number
}

export interface Stage2EngagementStatus {
  current_round: number
  fired_track_ids: number[]
  pending_track_ids: number[]
  confirmed_track_ids: number[]
  reengage_track_ids: number[]
  ready_to_close: boolean
  reason_codes: string[]
  updated_at: number
}

export interface Stage3FriendLink {
  balloon_track_id: number
  body_track_id: number
}

export interface Stage3EngagementStatus {
  current_round: number
  enemy_class: Stage3TargetClass | null
  enemy_balloon_track_id: number | null
  friend_links: Stage3FriendLink[]
  enemy_confirmation_state: string | null
  enemy_hit_confirmed: boolean
  friend_safety_verified: boolean
  friend_hit_suspected: boolean
  ready_to_close: boolean
  reason_codes: string[]
  shot_at: number | null
  updated_at: number
}

export interface Stage3RoundEvent {
  round_number: number
  enemy_class: Stage3TargetClass
  enemy_hit: boolean
  friend_hit: boolean
  points: number
  penalty: number
  miss_streak: number
  timestamp: number
}

export interface MissionState {
  active_stage: MissionStage
  elapsed_s: number
  timer_running: boolean
  stage1_hits: number
  stage1_wrong_hits: number
  stage1_order: string[]
  stage1_order_locked: boolean
  stage1_completed_targets: Stage1Target[]
  stage1_raw_points: number
  stage1_penalty_points: number
  stage1_events: Stage1Event[]
  stage2_round: number
  stage2_hits: number
  stage2_completed_rounds: number
  stage2_round_events: Stage2RoundEvent[]
  stage2_zero_hit_streak: number
  stage2_failed: boolean
  stage3_round: number
  stage3_hits: number
  stage3_friend_or_miss_penalties: number
  stage3_completed_rounds: number
  stage3_round_events: Stage3RoundEvent[]
  stage3_miss_streak: number
  stage3_failed: boolean
  updated_at: number
}

export interface MissionUpdate {
  active_stage?: MissionStage
  elapsed_s?: number
  timer_running?: boolean
}

export interface MissionScore {
  stage1_score: number
  stage2_score: number
  stage3_score: number
  active_score: number
  remaining_s: number
  total_estimated_score: number
  stage1_raw_points: number
  stage1_penalty_points: number
  stage1_bonus_points: number
  stage1_next_target: Stage1Target | null
  stage1_plan_locked: boolean
  stage2_round_scores: number[]
  stage2_zero_hit_streak: number
  stage2_failed: boolean
  stage2_passing_threshold_met: boolean
  stage3_round_scores: number[]
  stage3_miss_streak: number
  stage3_failed: boolean
  stage3_award_threshold_met: boolean
}

export interface MissionSnapshot {
  state: MissionState
  score: MissionScore
  no_physical_command_generated: boolean
}
